pkg/operator/operator.go

```diff
@@ -743,3 +776,244 @@ func (o *Operator) Config(ctx context.Context, key string) (*manifests.Config, e
 
 	return c, nil
 }
+
+// Upgradeable verifies whether the operator can be upgraded or not. It returns
+// the ConditionStatus with optional reason and message.
+func (o *Operator) Upgradeable(ctx context.Context) (configv1.ConditionStatus, string, string, error) {
+	if !o.lastKnowInfrastructureConfig.HighlyAvailableInfrastructure() {
+		return configv1.ConditionTrue, "", "", nil
+	}
+
+	workloadsCorrectlySpread, reason, message, err := o.WorkloadsCorrectlySpread(ctx)
+	if err != nil {
+		return configv1.ConditionUnknown, "", "", err
+	}
+
+	if !workloadsCorrectlySpread {
+		return configv1.ConditionFalse, reason, message, nil
+	}
+
+	return configv1.ConditionTrue, reason, message, nil
+}
+
+func (o *Operator) ensureNodesAreUncordonned(ctx context.Context) error {
+	for nodeName := range o.nodesCordon {
+		node, err := o.client.GetNode(ctx, nodeName)
+		if apierrors.IsNotFound(err) {
+			delete(o.nodesCordon, nodeName)
+			continue
+		}
+		if err != nil {
+			return err
+		}
+
+		err = drain.RunCordonOrUncordon(o.drainer, node, false)
+		if err != nil {
+			return err
+		}
+		delete(o.nodesCordon, node.Name)
+	}
+	return nil
+}
+
+func (o *Operator) spreadWorkloads(ctx context.Context, namespace string, sel map[string]string) (bool, error) {
+	podList, err := o.client.ListPods(ctx, namespace, metav1.ListOptions{LabelSelector: labels.FormatLabels(sel)})
+	if err != nil {
+		return false, err
+	}
+
+	if len(podList.Items) <= 1 {
+		return false, nil
+	}
+
+	node, err := o.client.GetNode(ctx, podList.Items[0].Spec.NodeName)
+	if err != nil {
+		return false, err
+	}
+
+	err = drain.RunCordonOrUncordon(o.drainer, node, true)
+	if err != nil {
+		return false, err
+	}
+	o.nodesCordon[node.Name] = struct{}{}
+
+	defer func() {
+		err = drain.RunCordonOrUncordon(o.drainer, node, false)
+		if err != nil {
+			klog.Errorf("Couldn't uncordon node %s: err %v", node.Name, err)
+		} else {
+			delete(o.nodesCordon, node.Name)
+		}
+	}()
+
+	var workloadSpread int
+	for _, pod := range podList.Items {
+		for _, vol := range pod.Spec.Volumes {
+			if vol.PersistentVolumeClaim != nil {
+				pvc, err := o.client.GetPersistentVolumeClaim(ctx, pod.Namespace, vol.PersistentVolumeClaim.ClaimName)
+				if apierrors.IsNotFound(err) {
+					continue
+				}
+				if err != nil {
+					return false, err
+				}
+
+				dropPVC, ok := pvc.Annotations[dropPVCAnnotation]
+				if !ok {
+					continue
+				}
+				if dropPVC == "yes" {
+					// Guard from deleting all PVCs to prevent complete data loss.
```

Comment:
for clarity, I would build the list of PVCs with the annotation first. Then remove annotations on all except the first one and proceed with the actual deletion. WDYT?

Reply:
I changed that and I am now building a list of PVCs and pods to be deleted. I then iterate over this list and give it to a `rebalanceWorkload` function that is responsible for condoning nodes and deleting PVCs/pods. The last element to be deleted is exempted if the annotation is set on all the PVCs of the workload.